flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogMigrationRestoreTarget.java

```diff
@@ -0,0 +1,223 @@
+
+    private final FunctionDelegationHelper functionDelegationHelper =
+            new FunctionDelegationHelper();
+
+    public ChangelogMigrationRestoreTarget(
+            AbstractKeyedStateBackend<K> keyedStateBackend,
+            ChangelogStateFactory changelogStateFactory) {
+        this.keyedStateBackend = keyedStateBackend;
+        this.changelogStateFactory = changelogStateFactory;
+    }
+
+    @Override
+    public KeyGroupRange getKeyGroupRange() {
+        return keyedStateBackend.getKeyGroupRange();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <N, S extends State, V> S createKeyedState(
+            TypeSerializer<N> namespaceSerializer, StateDescriptor<S, V> stateDescriptor)
+            throws Exception {
+        S keyedState =
+                keyedStateBackend.getOrCreateKeyedState(namespaceSerializer, stateDescriptor);
+        functionDelegationHelper.addOrUpdate(stateDescriptor);
+        final InternalKvState<K, N, V> kvState = (InternalKvState<K, N, V>) keyedState;
+        ChangelogState changelogState =
+                changelogStateFactory.create(stateDescriptor, kvState, keyedStateBackend);
+        return (S) changelogState;
+    }
+
+    @Nonnull
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends HeapPriorityQueueElement & PriorityComparable<? super T> & Keyed<?>>
+            KeyGroupedInternalPriorityQueue<T> createPqState(
+                    @Nonnull String stateName,
+                    @Nonnull TypeSerializer<T> byteOrderedElementSerializer) {
+        ChangelogKeyGroupedPriorityQueue<T> queue =
+                (ChangelogKeyGroupedPriorityQueue<T>)
+                        changelogStateFactory.getExistingState(
+                                stateName, StateMetaInfoSnapshot.BackendStateType.PRIORITY_QUEUE);
+        if (queue == null) {
+            KeyGroupedInternalPriorityQueue<T> internalPriorityQueue =
+                    keyedStateBackend.create(stateName, byteOrderedElementSerializer);
+            queue =
+                    changelogStateFactory.create(
+                            stateName, internalPriorityQueue, byteOrderedElementSerializer);
+        }
+        return queue;
+    }
+
+    @Override
+    public ChangelogState getExistingState(
+            String name, StateMetaInfoSnapshot.BackendStateType type) {
+        return changelogStateFactory.getExistingState(name, type);
+    }
+
+    @Override
+    public CheckpointableKeyedStateBackend<K> getRestoredKeyedStateBackend() {
+        // TODO: This inner class make the behaviour of the method of create consistent with
+        //  the method of getOrCreateKeyedState currently which could be removed
+        //  after we support state migration (in FLINK-23143).
+        //  It is also used to maintain FunctionDelegationHelper in the delegated state backend.
+        return new AbstractKeyedStateBackend<K>(keyedStateBackend) {
+
```

Comment:
I'm wondering whether it makes sense to extract this constructor call into static function; that would prevent leaking of `this` into the created object. 
WDYT?

Reply:
I think it's  a good suggestion. Done.